src/input/json_to_storage.cc

```diff
@@ -93,106 +231,362 @@ bool JSONPath::get_ref_from_head(string & ref_address)
-        tmp_str = address.substr(pos, new_pos - pos);
-        if (pos==0 && tmp_str == "") {
-            // absolute path
-            ref_path.go_to_root();
 
-        } else if ( ba::all( tmp_str, ba::is_digit()) ) {
-            // integer == index in array
-            if (ref_path.head() -> type() != json_spirit::array_type) {
-                THROW( ExcReferenceNotFound() << EI_RefAddress(*this) << EI_ErrorAddress(ref_path) << EI_RefStr(ref_address)
-                        << EI_Specification("there should be Array") );
-            }
 
-            if ( ref_path.down( atoi(tmp_str.c_str()) ) == NULL) {
-                THROW( ExcReferenceNotFound() << EI_RefAddress(*this) << EI_ErrorAddress(ref_path) << EI_RefStr(ref_address)
-                        << EI_Specification("index out of size of Array") );
-            }
+double PathJSON::get_double_value() const {
+    auto value_type = head()->type();
+    if (    value_type== json_spirit::real_type
+         || value_type == json_spirit::int_type) {
+        return head()->get_real();
+    } else {
+        THROW( JSONToStorage::ExcInputError() << JSONToStorage::EI_Specification("The value should be 'JSON real', but we found: ")
+                << JSONToStorage::EI_ErrorAddress(this) << JSONToStorage::EI_Format("JSON") << JSONToStorage::EI_JSON_Type( get_node_type() )
+        	 );
+    }
+	return 0.0;
+}
 
-        } else if (tmp_str == "..") {
-        	relative_ref = true;
-            if (ref_path.level() <= 0 ) {
-                THROW( ExcReferenceNotFound() << EI_RefAddress(*this) << EI_ErrorAddress(ref_path) << EI_RefStr(ref_address)
-                        << EI_Specification("can not go up from root") );
-            }
-            ref_path.up();
 
-        } else {
-            if (ref_path.head() -> type() != json_spirit::obj_type)
-                THROW( ExcReferenceNotFound() << EI_RefAddress(*this) << EI_ErrorAddress(ref_path) << EI_RefStr(ref_address)
-                        << EI_Specification("there should be Record") );
-            if ( ref_path.down(tmp_str) == NULL )
-                THROW( ExcReferenceNotFound() << EI_RefAddress(*this) << EI_ErrorAddress(ref_path) << EI_RefStr(ref_address)
-                        << EI_Specification("key '"+tmp_str+"' not found") );
-        }
-        pos = new_pos+1;
+
+std::string PathJSON::get_string_value() const {
+    if (head()->type() == json_spirit::str_type) {
+        return head()->get_str();
+    } else {
+        THROW( JSONToStorage::ExcInputError() << JSONToStorage::EI_Specification("The value should be 'JSON string', but we found: ")
+                << JSONToStorage::EI_ErrorAddress(this) << JSONToStorage::EI_Format("JSON") << JSONToStorage::EI_JSON_Type( get_node_type() )
+             );
     }
-    if (relative_ref) {
-    	xprintf(Msg, "Referencing '%s' to '%s'.\n", this->str().c_str(), ref_path.str().c_str());
+	return "";
+}
+
+
+
+std::string PathJSON::get_node_type() const {
```

Comment:
Use same implementation as in PathYAML, or at least make json_type_names a local variable in this method.